// @trimbleinc/modus-react-bootstrap/cjs/ButtonGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _classnames = _interopRequireDefault(require("classnames"));
var _react = _interopRequireDefault(require("react"));
var _ThemeProvider = require("./ThemeProvider");
var _jsxRuntime = require("react/jsx-runtime");
const defaultProps = {
  vertical: false,
  toggle: false,
  role: 'group'
};
const ButtonGroup = /*#__PURE__*/_react.default.forwardRef(({
  bsPrefix,
  size,
  toggle,
  vertical,
  className,
  // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
  as: Component = 'div',
  ...rest
}, ref) => {
  const prefix = (0, _ThemeProvider.useBootstrapPrefix)(bsPrefix, 'btn-group');
  let baseClass = prefix;
  if (vertical) baseClass = `${prefix}-vertical`;
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(Component, {
    ...rest,
    ref: ref,
    className: (0, _classnames.default)(className, baseClass, size && `${prefix}-${size}`, toggle && `${prefix}-toggle`)
  });
});
ButtonGroup.displayName = 'ButtonGroup';
ButtonGroup.defaultProps = defaultProps;
var _default = ButtonGroup;
exports.default = _default;
module.exports = exports.default;
