// @trimbleinc/modus-react-bootstrap/cjs/Button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _classnames = _interopRequireDefault(require("classnames"));
var _react = _interopRequireDefault(require("react"));
var _ThemeProvider = require("./ThemeProvider");
var _SafeAnchor = _interopRequireDefault(require("./SafeAnchor"));
var _jsxRuntime = require("react/jsx-runtime");
const defaultProps = {
  variant: 'primary',
  active: false,
  disabled: false
};
const Button = /*#__PURE__*/_react.default.forwardRef(({
  bsPrefix,
  variant,
  size,
  active,
  className,
  block,
  type,
  as,
  ...props
}, ref) => {
  const prefix = (0, _ThemeProvider.useBootstrapPrefix)(bsPrefix, 'btn');
  const classes = (0, _classnames.default)(className, prefix, active && 'active', variant && `${prefix}-${variant}`, block && `${prefix}-block`, size && `${prefix}-${size}`);
  if (props.href) {
    return /*#__PURE__*/(0, _jsxRuntime.jsx)(_SafeAnchor.default, {
      ...props,
      as: as,
      ref: ref,
      className: (0, _classnames.default)(classes, props.disabled && 'disabled')
    });
  }
  if (ref) {
    props.ref = ref;
  }
  if (type) {
    props.type = type;
  } else if (!as) {
    props.type = 'button';
  }
  const Component = as || 'button';
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(Component, {
    ...props,
    className: classes
  });
});
Button.displayName = 'Button';
Button.defaultProps = defaultProps;
var _default = Button;
exports.default = _default;
module.exports = exports.default;
