// @trimbleinc/modus-react-bootstrap/cjs/DataTable.props.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.DataTablePropTypes = void 0;
var _propTypes = _interopRequireDefault(require("prop-types"));
/*!
  Modus React Bootstrap 
  A React-based component library developed as a common, open source platform for all of Trimble’s web applications built on React.
  Extends React-Bootstrap v1.6.5
  Copyright (c) 2022 Trimble Inc.
 */

const DataTablePropTypes = {
  /** A HTML id attribute, necessary for proper form accessibility. */
  id: _propTypes.default.string.isRequired,
  /**
   * An array of column data of type `DataTableColumn` for the Table Headers.
   */
  columns: _propTypes.default.array.isRequired,
  /**
   * Default page size if pagination is enabled.
   */
  pageSize: _propTypes.default.number,
  /**
   * An array of page size options if pagination is enabled.
   */
  pageSizeOptions: _propTypes.default.arrayOf(_propTypes.default.number),
  /**
   * An array of data for the Table Rows.
   */
  data: _propTypes.default.array.isRequired,
  /**
   * Enables column resizing.
   */
  resizeColumns: _propTypes.default.bool,
  /**
   * Enables checkbox to select the rows.
   */
  checkBoxRowSelection: _propTypes.default.bool,
  /**
   * Enables multiple row selection.
   */
  multipleRowSelection: _propTypes.default.bool,
  /**
   * Disables on-click row selection.
   */
  disableRowSelectionOnClick: _propTypes.default.bool,
  /**
   * Displays all the rows and hides pagination panel.
   */
  disablePagination: _propTypes.default.bool,
  /**
   * Removes sorting icons from the column header.
   */
  disableSorting: _propTypes.default.bool,
  /**
   * Removes filter panel.
   */
  disableFiltering: _propTypes.default.bool,
  /**
   * Disables dragging function on Table headers.
   */
  disableDragging: _propTypes.default.bool,
  /**
   * Adds zebra-striping to any table row within the `<tbody>`.
   */
  striped: _propTypes.default.bool,
  /**
   * Adds borders on all sides of the table and cells.
   */
  bordered: _propTypes.default.bool,
  /**
   * Removes all borders on the table and cells, including table header.
   */
  borderless: _propTypes.default.bool,
  /**
   * Enable a hover state on table rows within a `<tbody>`.
   */
  hover: _propTypes.default.bool,
  /**
   * Make tables more compact by cutting cell padding in half by setting
   * size as `sm`.
   */
  size: _propTypes.default.string,
  /**
   * Pins the first column.
   */
  stickyFirstColumn: _propTypes.default.bool,
  /**
   * Invert the colors of the table — with light text on dark backgrounds
   * by setting variant as `dark`.
   */
  variant: _propTypes.default.string,
  /**
   * Responsive tables allow tables to be scrolled horizontally with ease.
   * Across every breakpoint, use `responsive` for horizontally
   * scrolling tables. Responsive tables are wrapped automatically in a `div`.
   * Use `responsive="sm"`, `responsive="md"`, `responsive="lg"`, or
   * `responsive="xl"` as needed to create responsive tables up to
   * a particular breakpoint. From that breakpoint and up, the table will
   * behave normally and not scroll horizontally.
   */
  responsive: _propTypes.default.oneOfType([_propTypes.default.bool, _propTypes.default.string]),
  /**
   * Callback when rows selected or unselected.
   *
   * ```js
   * function onRowSelectionChange (rows: any[]) => void
   * ```
   */
  onRowSelectionChange: _propTypes.default.func,
  /**
   * Function that returns a custom filter panel.
   *
   * ```js
   * function filterPanel(props: DataTableFilterPanelProps) => React.ReactElement
   * params:
   * - props:
   *    filterColumns - An array of filter columns (), `DataTableColumnInstance[]`
   *    activeFilters - An array of active filters
   *    resetFilter - Function to reset a column's filter
   *    resetAllFilters - Function to reset all the filters
   *    globalFilter - Global filter value
   *    setGlobalFilter - Function to set the global filter value
   * ```
   */
  filterPanel: _propTypes.default.func,
  /**
   * Function that returns a template for Dragging item.
   *
   * ```js
   * function dragTemplate(column: DataTableColumnInstance) => React.ReactElement
   *  column - An instance of the column object that is dragged
   * ```
   */
  dragTemplate: _propTypes.default.func
};
exports.DataTablePropTypes = DataTablePropTypes;
