// @trimbleinc/modus-react-bootstrap/cjs/NavLink.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _classnames = _interopRequireDefault(require("classnames"));
var _react = _interopRequireDefault(require("react"));
var _SafeAnchor = _interopRequireDefault(require("./SafeAnchor"));
var _AbstractNavItem = _interopRequireDefault(require("./AbstractNavItem"));
var _ThemeProvider = require("./ThemeProvider");
var _jsxRuntime = require("react/jsx-runtime");
const defaultProps = {
  disabled: false,
  as: _SafeAnchor.default
};
const NavLink = /*#__PURE__*/_react.default.forwardRef(({
  bsPrefix,
  disabled,
  className,
  href,
  eventKey,
  onSelect,
  as,
  ...props
}, ref) => {
  bsPrefix = (0, _ThemeProvider.useBootstrapPrefix)(bsPrefix, 'nav-link');
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_AbstractNavItem.default, {
    ...props,
    href: href,
    ref: ref,
    eventKey: eventKey,
    as: as,
    disabled: disabled,
    onSelect: onSelect,
    className: (0, _classnames.default)(className, bsPrefix, disabled && 'disabled')
  });
});
NavLink.displayName = 'NavLink';
NavLink.defaultProps = defaultProps;
var _default = NavLink;
exports.default = _default;
module.exports = exports.default;
