// @trimbleinc/modus-react-bootstrap/cjs/Popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.default = void 0;
var _classnames = _interopRequireDefault(require("classnames"));
var _react = _interopRequireDefault(require("react"));
var _isRequiredForA11y = _interopRequireDefault(require("prop-types-extra/lib/isRequiredForA11y"));
var _ThemeProvider = require("./ThemeProvider");
var _PopoverTitle = _interopRequireDefault(require("./PopoverTitle"));
var _PopoverContent = _interopRequireDefault(require("./PopoverContent"));
var _jsxRuntime = require("react/jsx-runtime");
const defaultProps = {
  placement: 'right'
};
const Popover = /*#__PURE__*/_react.default.forwardRef(({
  bsPrefix,
  placement,
  className,
  style,
  children,
  content,
  arrowProps,
  popper: _,
  show: _1,
  ...props
}, ref) => {
  const decoratedBsPrefix = (0, _ThemeProvider.useBootstrapPrefix)(bsPrefix, 'popover');
  const [primaryPlacement] = (placement == null ? void 0 : placement.split('-')) || [];
  return /*#__PURE__*/(0, _jsxRuntime.jsxs)("div", {
    ref: ref,
    role: "tooltip",
    style: style,
    "x-placement": primaryPlacement,
    className: (0, _classnames.default)(className, decoratedBsPrefix, primaryPlacement && `bs-popover-${primaryPlacement}`),
    ...props,
    children: [/*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
      className: "arrow",
      ...arrowProps
    }), content ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_PopoverContent.default, {
      children: children
    }) : children]
  });
});
Popover.defaultProps = defaultProps;
Popover.Title = _PopoverTitle.default;
Popover.Content = _PopoverContent.default;
var _default = Popover;
exports.default = _default;
module.exports = exports.default;
