// @trimbleinc/modus-react-bootstrap/esm/DataTable.js
/*!
  Modus React Bootstrap 
  A React-based component library developed as a common, open source platform for all of Trimble’s web applications built on React.
  Extends React-Bootstrap v1.6.5
  Copyright (c) 2022 Trimble Inc.
 */

import React, { useCallback, useEffect, useRef } from 'react';
import { useSortBy, usePagination, useResizeColumns, useFlexLayout, useRowSelect, useFilters, useGlobalFilter, useColumnOrder } from 'react-table';
import classNames from 'classnames';
import { DATATABLE_DEFAULT_PAGE_SIZES, checkBoxSelectionHook, getCellStyles, stateReducer } from './DataTableHelpers';
import Table from './Table';
import TablePagination from './TablePagination';
import DataTableHeaderCell from './DataTableHeaderCell';
import DataTableDragdropProvider from './DataTableDragdropProvider';
import DataTableContextMenuProvider from './DataTableContextMenuProvider';
import useDataTableInstance from './useDataTableInstance';
import { DataTablePropTypes } from './DataTable.props';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
function RenderFilterPanel(props, tableInstance) {
  const {
    filterPanel,
    disableFiltering
  } = props;
  const {
    filterColumns,
    resetFilter,
    resetAllFilters,
    setGlobalFilter,
    state: {
      filters: activeFilters,
      globalFilter
    }
  } = tableInstance;
  return filterPanel && !disableFiltering ? filterPanel({
    filterColumns,
    activeFilters,
    resetFilter,
    resetAllFilters,
    globalFilter,
    setGlobalFilter
  }) : null;
}
function RenderPagination(props, tableInstance) {
  const {
    data,
    disablePagination,
    pageSizeOptions,
    size
  } = props;
  const {
    gotoPage,
    setPageSize,
    state: {
      pageIndex,
      pageSize
    }
  } = tableInstance;
  return !disablePagination ? /*#__PURE__*/_jsx(TablePagination, {
    count: data.length,
    pageIndex: pageIndex,
    pageSize: pageSize,
    onPageChange: gotoPage,
    pageSizeOptions: pageSizeOptions || DATATABLE_DEFAULT_PAGE_SIZES,
    onPageSizeChange: setPageSize,
    size: size
  }) : null;
}
function RenderTableHeader(props, tableInstance) {
  const {
    size,
    dragTemplate
  } = props;
  const {
    headerGroups,
    allColumns,
    visibleColumns,
    getAllHeadersInAGroup,
    toggleHideColumn,
    toggleHideAllColumns,
    setColumnOrder
  } = tableInstance;
  return /*#__PURE__*/_jsx("thead", {
    className: "bg-gray-light",
    children: headerGroups.map(headerGroup => /*#__PURE__*/_jsx("tr", {
      ...headerGroup.getHeaderGroupProps(),
      role: "row",
      children: /*#__PURE__*/_jsx(DataTableContextMenuProvider, {
        size: size,
        allColumns: allColumns,
        toggleHideAllColumns: toggleHideAllColumns,
        toggleHideColumn: toggleHideColumn,
        children: /*#__PURE__*/_jsx(DataTableDragdropProvider, {
          visibleColumns: visibleColumns,
          setColumnOrder: setColumnOrder,
          dragItemTemplate: dragTemplate,
          children: getAllHeadersInAGroup(headerGroup.headers, headerGroup.id).map((column, index) => /*#__PURE__*/_jsx(DataTableHeaderCell, {
            header: column,
            onToggleHideColumn: toggleHideColumn,
            role: "columnheader",
            "aria-sort": column.isSorted && (column.isSortedDesc ? 'descending' : 'ascending') || 'none',
            className: "bg-gray-light sticky-top",
            "aria-colindex": index + 1,
            children: column.render('Header')
          }, column.id))
        })
      })
    }))
  });
}
function RenderTableBody(props, tableInstance, handleRowClickFn) {
  const {
    checkBoxRowSelection,
    disablePagination
  } = props;
  const {
    rows,
    page,
    prepareRow,
    getTableBodyProps
  } = tableInstance;
  return /*#__PURE__*/_jsx("tbody", {
    ...getTableBodyProps(),
    children: (disablePagination ? rows : page).map(row => {
      prepareRow(row);
      return /*#__PURE__*/_jsx("tr", {
        ...row.getRowProps(),
        onClick: () => handleRowClickFn(row),
        className: classNames(row.isSelected && 'selected'),
        role: "row",
        children: row.cells.map((cell, index) => /*#__PURE__*/_jsx("td", {
          ...cell.getCellProps(getCellStyles),
          className: classNames(index === 0 && `${checkBoxRowSelection && 'icon-only checkbox-selector-cell'}`),
          role: "cell",
          "aria-rowindex": index + 1,
          children: cell.render('Cell')
        }))
      });
    })
  });
}
function RenderTable(props, tableInstance, handleRowClickFn) {
  const {
    data,
    striped,
    stickyFirstColumn,
    bordered,
    borderless,
    hover,
    size,
    variant,
    responsive
  } = props;
  const {
    allColumns,
    isResizing,
    getTableProps
  } = tableInstance;
  return /*#__PURE__*/_jsxs(Table, {
    striped: striped,
    bordered: bordered,
    borderless: borderless,
    hover: hover,
    size: size,
    variant: variant,
    responsive: responsive,
    ...getTableProps(),
    role: "table",
    "aria-colcount": allColumns.length,
    "aria-rowcount": data.length,
    className: classNames(stickyFirstColumn && 'table-sticky-first-column', isResizing && 'resizing'),
    children: [RenderTableHeader(props, tableInstance), RenderTableBody(props, tableInstance, handleRowClickFn)]
  });
}
function DataTable(props) {
  /* eslint-disable */
  const {
    id,
    columns,
    pageSize: pageSizeProp,
    pageSizeOptions,
    data,
    resizeColumns,
    checkBoxRowSelection,
    multipleRowSelection,
    disableRowSelectionOnClick,
    disablePagination,
    disableSorting,
    disableFiltering,
    disableDragging,
    striped,
    stickyFirstColumn,
    bordered,
    borderless,
    hover,
    size,
    variant,
    responsive,
    onRowSelectionChange,
    ref,
    className,
    filterPanel,
    dragTemplate,
    ...htmlProps
  } = props;
  /* eslint-enable */

  const resolvedRef = useRef(null) || ref;
  const enableRowSelection = !disableRowSelectionOnClick || checkBoxRowSelection;
  const disableMouseClickRowSelection = disableRowSelectionOnClick || checkBoxRowSelection;

  // Conditional Table Hooks Array for react-table
  const conditionalHooks = [useFlexLayout];
  if (filterPanel && !disableFiltering) conditionalHooks.push(useFilters, useGlobalFilter);
  if (!disableSorting) conditionalHooks.push(useSortBy);
  if (!disablePagination) conditionalHooks.push(usePagination);
  if (resizeColumns) conditionalHooks.push(useResizeColumns);
  if (enableRowSelection) conditionalHooks.push(useRowSelect);
  if (!disableDragging) conditionalHooks.push(useColumnOrder);
  if (checkBoxRowSelection && !columns.find(col => col.accessor === 'selector')) {
    conditionalHooks.push(hooks => checkBoxSelectionHook(hooks, id, multipleRowSelection));
  }
  const tableOptions = {
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    initialState: {
      pageIndex: 0,
      pageSize: pageSizeProp || 10
    },
    ...(!multipleRowSelection && stateReducer)
  };
  const tableInstance = useDataTableInstance(columns, data, tableOptions, conditionalHooks);
  const {
    selectedFlatRows
  } = tableInstance;

  // Callback APIs
  useEffect(() => {
    if (onRowSelectionChange) onRowSelectionChange(selectedFlatRows.map(d => d.original));
  }, [selectedFlatRows, onRowSelectionChange]);
  const handleRowClick = useCallback(row => {
    if (!disableMouseClickRowSelection) {
      row.toggleRowSelected(!row.isSelected);
    }
  }, [disableMouseClickRowSelection]);
  return /*#__PURE__*/_jsx("div", {
    ref: resolvedRef,
    className: classNames('mrb-data-table', bordered && 'mrb-data-table-bordered', className),
    ...htmlProps,
    children: /*#__PURE__*/_jsxs("div", {
      className: classNames('d-flex flex-column overflow-hidden'),
      children: [RenderFilterPanel(props, tableInstance), /*#__PURE__*/_jsxs("div", {
        className: classNames('mrb-data-table-container d-flex flex-column overflow-hidden'),
        children: [
        /*#__PURE__*/
        // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
        _jsx("div", {
          className: classNames('scrollable'),
          tabIndex: 0,
          children: RenderTable(props, tableInstance, handleRowClick)
        }), RenderPagination(props, tableInstance)]
      })]
    })
  });
}
export default DataTable;
