// @trimbleinc/modus-react-bootstrap/esm/FormFileInput.js
import classNames from 'classnames';
import React, { useContext } from 'react';
import FormContext from './FormContext';
import { useBootstrapPrefix } from './ThemeProvider';
import { jsx as _jsx } from "react/jsx-runtime";
const FormFileInput = /*#__PURE__*/React.forwardRef(({
  id,
  bsPrefix,
  bsCustomPrefix,
  className,
  isValid,
  isInvalid,
  lang,
  // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
  as: Component = 'input',
  ...props
}, ref) => {
  const {
    controlId,
    custom
  } = useContext(FormContext);
  const type = 'file';
  const [prefix, defaultPrefix] = custom ? [bsCustomPrefix, 'custom-file-input'] : [bsPrefix, 'form-control-file'];
  bsPrefix = useBootstrapPrefix(prefix, defaultPrefix);
  return /*#__PURE__*/_jsx(Component, {
    ...props,
    ref: ref,
    id: id || controlId,
    type: type,
    lang: lang,
    className: classNames(className, bsPrefix, isValid && 'is-valid', isInvalid && 'is-invalid')
  });
});
FormFileInput.displayName = 'FormFileInput';
export default FormFileInput;
