// @trimbleinc/modus-react-bootstrap/esm/Message.js
/*!
  Modus React Bootstrap 
  A React-based component library developed as a common, open source platform for all of Trimble’s web applications built on React.
  Extends React-Bootstrap v1.6.5
  Copyright (c) 2022 Trimble Inc.
 */
import * as React from 'react';
import { forwardRef } from 'react';
import classNames from 'classnames';
import * as PropTypes from 'prop-types';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const propTypes = {
  /**
   * The visual variant
   *
   * @type {'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'dark' | 'light'}
   */
  variant: PropTypes.string,
  /**
   * message text
   *
   */
  label: PropTypes.string,
  /**
   * Icon must be either a valid JSX element or a modus icon, ex: 'info_outlined'. Icon: `ReactElement | string`
   *
   */
  icon: PropTypes.element,
  /**
   * Controls the visual state.
   *
   * @controllable onClose
   */
  show: PropTypes.bool
};
const defaultProps = {
  show: true
};
const prefix = 'message';
const Message = /*#__PURE__*/forwardRef(({
  variant,
  message,
  icon,
  show,
  className,
  ...props
}, ref) => {
  const iconElement = typeof icon === 'string' ? /*#__PURE__*/_jsx("i", {
    className: "modus-icons",
    "aria-hidden": "true",
    children: icon
  }) : icon;
  const msg = /*#__PURE__*/_jsxs("div", {
    ...props,
    ref: ref,
    className: classNames(prefix, variant && `${prefix}-${variant}`, className),
    children: [iconElement, message, props.children]
  });
  return show ? msg : null;
});
Message.displayName = 'Message';
Message.defaultProps = defaultProps;
Message.propTypes = propTypes;
export default Message;
