// @trimbleinc/modus-react-bootstrap/esm/ModalHeader.js
import classNames from 'classnames';
import React, { useContext } from 'react';
import useEventCallback from '@restart/hooks/useEventCallback';
import { useBootstrapPrefix } from './ThemeProvider';
import CloseButton from './CloseButton';
import ModalContext from './ModalContext';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
const defaultProps = {
  closeLabel: 'Close',
  closeButton: false
};
const ModalHeader = /*#__PURE__*/React.forwardRef(({
  bsPrefix,
  closeLabel,
  closeButton,
  onHide,
  className,
  children,
  ...props
}, ref) => {
  bsPrefix = useBootstrapPrefix(bsPrefix, 'modal-header');
  const context = useContext(ModalContext);
  const handleClick = useEventCallback(() => {
    if (context) context.onHide();
    if (onHide) onHide();
  });
  return /*#__PURE__*/_jsxs("div", {
    ref: ref,
    ...props,
    className: classNames(className, bsPrefix),
    children: [children, closeButton && /*#__PURE__*/_jsx(CloseButton, {
      label: closeLabel,
      onClick: handleClick
    })]
  });
});
ModalHeader.displayName = 'ModalHeader';
ModalHeader.defaultProps = defaultProps;
export default ModalHeader;
